src/Evaluator.hs

```diff
@@ -1,20 +1,77 @@
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE BangPatterns #-}
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE GADTSyntax #-}
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE ParallelListComp #-}
 {-# LANGUAGE RecordWildCards #-}
 {-# LANGUAGE TemplateHaskell #-}
-module Evaluator where
+
+{- Note [The CEK interpreter]:
+
+The Klister interpreter is a straightforward implementation of a CEK
+interpreter. The interpreter keeps three kinds of state:
+
+-- C: Control      ::= The thing that is being evaluated
+-- E: Environment  ::= The interpreter environment
+-- K: Kontinuation ::= The syntactic context of the thing that is being interpreted
+
+Why a CEK? A CEK interpreter allows us to have precise control over the
+evaluation of a klister program. For example, because the interpreter keeps a
+reference to the kontinuation we can provide stack traces. This handle also
+makes a more advanced debugger possible. Upon an evaluation error we could save
+the kontinuation stack, over write a variable in the environment a la common
+lisp or even rewind the evaluation
+
+See Matthias Felleison's course website for a good reference:
+https://felleisen.org/matthias/4400-s20/lecture23.html
+
+The bird's eye view:
+
+The evaluator crawl's the input AST and progresses in three modes:
+
+-- 'Down': meaning that the evaluator is searching for a redex to evaluate and
+-- therefore moving "down" the AST.
+
+-- 'Up': meaning that the evaluator has evaluated some redex to a value and is
+-- passing that value "up" the execution stack.
+
+-- 'Er': meaning that something has gone wrong, the stack is captured and the Er
+-- will float up to be handled by the caller of the evaluator.
+
+All interesting things happen by matching on 'Kont', the continuation. This
+allows the evaluator to know exactly what needs to happen in order to continue.
+
+-- TODO: #108 describe the how the debugger hooks in
+
+-}
+
+module Evaluator
+  ( EvalError (..)
+  , EvalResult (..)
+  , TypeError (..)
+  , evaluate
+  , evaluateIn
+  , evaluateWithExtendedEnv
+  , evalErrorType
+  , evalErrorText
```

Comment:
`evalErrorText` is not used anywhere? that's strange, where do the errors get printed? Maybe the `Show` instance is used instead of `evalErrorText` by mistake?